lib/templatesimplifier.h

```diff
@@ -66,179 +63,174 @@ class CPPCHECKLIB TemplateSimplifier {
-                                     std::list<TokenAndName> &templateInstantiations);
+    void replaceTemplateUsage(Token *const instantiationToken,
+                              const std::string &templateName,
+                              const std::list<std::string> &typeStringsUsedInTemplateInstantiation,
+                              const std::string &newName,
+                              const std::vector<const Token *> &typesUsedInTemplateInstantiation);
 
     /**
-     * Simplify templates
-     * @param tokenlist token list
-     * @param errorlogger error logger
-     * @param mSettings settings
-     * @param maxtime time when the simplification should be stopped
-     * @param codeWithTemplates output parameter that is set if code contains templates
-     */
-    static void simplifyTemplates(
-        TokenList& tokenlist,
-        ErrorLogger* errorlogger,
-        const Settings *mSettings,
-        const std::time_t maxtime,
-        bool &codeWithTemplates);
-
-    /**
-     * Simplify constant calculations such as "1+2" => "3"
-     * @param tok start token
-     * @return true if modifications to token-list are done.
-     *         false if no modifications are done.
+     * Expand specialized templates : "template<>.."
+     * @return names of expanded templates
      */
-    static bool simplifyNumericCalculations(Token *tok);
+    std::set<std::string> expandSpecialized();
 
     /**
-     * Simplify constant calculations such as "1+2" => "3".
-     * This also performs simple cleanup of parentheses etc.
-     * @param _tokens start token
-     * @return true if modifications to token-list are done.
-     *         false if no modifications are done.
+     * @brief TemplateParametersInDeclaration
+     * @param tok  template < typename T, typename S >
+     *                        ^ tok
+     * @param typeParametersInDeclaration  template < typename T, typename S >
+     *                                                         ^ [0]       ^ [1]
+     * @return  template < typename T, typename S >
+     *                                              ^ return
      */
-    static bool simplifyCalculations(Token *_tokens);
+    const Token * getTemplateParametersInDeclaration(
+        const Token * tok,
+        std::vector<const Token *> & typeParametersInDeclaration);
 
-private:
     /**
      * Remove a specific "template < ..." template class/function
      */
-    static bool removeTemplate(Token *tok);
+    bool removeTemplate(Token *tok);
 
     /** Syntax error */
     static void syntaxError(const Token *tok);
 
+    bool matchSpecialization(
+        const Token *templateDeclarationNameToken,
+        const Token *templateInstantiationNameToken,
+        const std::list<const Token *> & specializations);
+
+    void eraseTokens(Token *begin, const Token *end);
+    void deleteThis(Token *tok);
+
+    Tokenizer *mTokenizer;
```

Comment:
We only use the `TokenList` as far as I see.

Reply:
Fixed